pyanaconda/modules/payload/base/handler_base.py

```diff
@@ -19,15 +19,71 @@
 #
 from abc import ABCMeta, abstractmethod
 
+from pyanaconda.core.signal import Signal
+from pyanaconda.modules.common.errors.payload import IncompatibleSourceError
 from pyanaconda.modules.common.base import KickstartBaseModule
 
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
 
 class PayloadHandlerBase(KickstartBaseModule, metaclass=ABCMeta):
     """Base class for all the payload handler modules.
 
     This will contain all API specific to payload handlers which will be called
     by the base payload module.
     """
+    def __init__(self):
+        super().__init__()
+        self._sources = set()
+        self.sources_changed = Signal()
+
+    @property
+    @abstractmethod
+    def supported_source_kinds(self):
+        """Get list of supported source types.
+
+        :return: list of supported source types
+        :rtype: [values from payload.base.constants.SourceType]
+        """
+        pass
+
+    @property
+    def sources(self):
+        """Get list of sources attached to this payload handler.
+
+        :return: list of source objects attached to this handler
+        :rtype: [instance of PayloadSourceBase class]
+        """
+        return self._sources
+
+    def add_source(self, source):
+        """Add source to the list of sources.
+
+        :param source: source object
+        :type source: instance of pyanaconda.modules.payload.base.source_base.PayloadSourceBase
+        :raises: IncompatibleSourceError
+        """
+        # TODO: Add test for this when there will be public API
+        if source.kind not in self.supported_source_kinds:
+            raise IncompatibleSourceError("Source type {} is not supported by this handler."
+                                          .format(source.kind))
+
```

Comment:
I wonder why you are not using something like
```
if source not in self._sources:
    self._sources.add(source)
    log.debug("New source %s was added.", source.kind)
    self.sources_changed.emit()
```
which seems more natural to me, but I don't insist on using it.

Reply:
Good question and unfortunately don't know. I guess I had a blackout or what. I'll change this part.

Reply:
Fixed.